src/hotspot/share/cds/dynamicArchive.cpp

```diff
@@ -349,6 +396,77 @@ class VM_PopulateDynamicDumpSharedSpace: public VM_GC_Sync_Operation {
   }
 };
 
+GrowableArray<ObjArrayKlass*>* DynamicArchive::_array_klasses = nullptr;
+Array<ObjArrayKlass*>* DynamicArchive::_dynamic_archive_array_klasses = nullptr;
+
+void DynamicArchive::append_array_klass(ObjArrayKlass* ak) {
+  if (_array_klasses == nullptr) {
+    _array_klasses = new (mtClassShared) GrowableArray<ObjArrayKlass*>(50, mtClassShared);
+  }
+  _array_klasses->append(ak);
+}
+
+void DynamicArchive::dump_array_klasses() {
+  assert(DynamicDumpSharedSpaces, "DynamicDumpSharedSpaces only");
+  if (_array_klasses != nullptr) {
+    ArchiveBuilder* builder = ArchiveBuilder::current();
+    int num_array_klasses = _array_klasses->length();
+    _dynamic_archive_array_klasses =
+        ArchiveBuilder::new_ro_array<ObjArrayKlass*>(num_array_klasses);
+    for (int i = 0; i < num_array_klasses; i++) {
+      ObjArrayKlass* oak = _array_klasses->at(i);
+      _dynamic_archive_array_klasses->at_put(i, oak);
+      builder->write_pointer_in_buffer(_dynamic_archive_array_klasses->adr_at(i), oak);
+    }
+  }
+}
+
+void DynamicArchive::setup_array_klasses() {
+  if (_dynamic_archive_array_klasses != nullptr) {
+    for (int i = 0; i < _dynamic_archive_array_klasses->length(); i++) {
+      ObjArrayKlass* oak = _dynamic_archive_array_klasses->at(i);
+      assert(!oak->is_typeArray_klass(), "all type array classes must be in static archive");
+
+      Klass* elm = oak->element_klass();
+      assert(MetaspaceShared::is_shared_static((void*)elm), "must be");
+
+      if (elm->is_instance_klass()) {
+        assert(InstanceKlass::cast(elm)->array_klasses() == nullptr, "must be");
+        InstanceKlass::cast(elm)->release_set_array_klasses(oak);
```

Comment:
I know you were asked to use this here but a release operation only has meaning when paired with an acquire - so where is the matching acquire?

Reply:
I don't know if this code is run at a safepoint, but because of this bug https://bugs.openjdk.org/browse/JDK-8308745, we can't use a lock to allocate objArrayKlass, so we'll likely need the matching acquire when we fix that.

Reply:
The code in question is being called during VM startup.
```
DynamicArchive::setup_array_klasses() : void
	MetaspaceShared::initialize_shared_spaces() : void
		universe_init() : jint
			init_globals() : jint
				Threads::create_vm(JavaVMInitArgs *, bool *) : jint
					JNI_CreateJavaVM_inner(JavaVM * *, void * *, void *) : jint
						JNI_CreateJavaVM(JavaVM * *, void * *, void *) : jint
```
Maybe it is safe not to use the "release_" version for setting `array_klasses`?

Reply:
The VM is still single-threaded at that point so no "release" is necessary in this code.

Reply:
Thanks. There is no `set_array_klasses()` function yet. So I added one to instanceKlass.hpp in my latest commit https://github.com/openjdk/jdk/pull/14959/commits/6e69ba63cb01d0799d5609a0e72349d3659ef239.